File: src/lib/dashboard-converter.ts
import {assignmentString, block, blockList, convertFromDefinition} from "./utils.js";

const DASHBOARD = {
    description: (v: object) => assignmentString("description", v),
    id: (_: any) => "",
    is_read_only: (v: object) => assignmentString("is_read_only", v),
    layout_type: (v: object) => assignmentString("layout_type", v),
    notify_list: (v: object) => assignmentString("notify_list", v),
    template_variable_presets: (v: object[]) => blockList(v, "template_variable_preset", (k1, v1) => convertFromDefinition(TEMPLATE_VARIABLE_PRESET, k1, v1)),
    template_variables: (v: object[]) => blockList(v, "template_variable", assignmentString),
    title: (v: object) => assignmentString("title", v),
    url: (v: object) => assignmentString("url", v),
    widgets: (v: object[]) => convertWidgets(v),
};

const WIDGET = {
    definition: (v: object) => widgetDefinition(v),
    id: (_: any) => "",
    layout: (v: object) => block("widget_layout", v, assignmentString),
};

const TEMPLATE_VARIABLE_PRESET = {
    name: (v: object) => assignmentString("name", v),
    template_variables: (v: object[]) => blockList(v, "template_variable", assignmentString),
};

const WIDGET_DEFINITION = {
    alert_id: (v: object) => assignmentString("alert_id", v),
    autoscale: (v: object) => assignmentString("autoscale", v),
    background_color: (v: object) => assignmentString("background_color", v),
    check: (v: object) => assignmentString("check", v),
    color: (v: object) => assignmentString("color", v),
    color_by_groups: (v: object) => assignmentString("color_by_groups", v),
    color_preference: (v: object) => assignmentString("color_preference", v),
    columns: (v: object) => assignmentString("columns", v),
    content: (v: object) => assignmentString("content", v),
    count: (_: any) => "", // 2.23.0 deprecated, see docs for widget.manage_status_definition
    custom_links: (v: object[]) => blockList(v, "custom_link", assignmentString),
    custom_unit: (v: object) => assignmentString("custom_unit", v),
    display_format: (v: object) => assignmentString("display_format", v),
    env: (v: object) => assignmentString("env", v),
    event: (v: object) => block("event", v, assignmentString),
    event_size: (v: object) => assignmentString("event_size", v),
    filters: (v: object) => assignmentString("filters", v),
    font_size: (v: object) => assignmentString("font_size", v),
    global_time_target: (_: any) => "",
    group: (v: object) => assignmentString("group", v),
    group_by: (v: object) => assignmentString("group_by", v),
    grouping: (v: object) => assignmentString("grouping", v),
    has_padding: (_: any) => "", // 2.23.0 not described in docs, occurs in widget.note_definition json
    has_search_bar: (v: object) => assignmentString("has_search_bar", v),
    hide_zero_counts: (v: object) => assignmentString("hide_zero_counts", v),
    indexes: (v: object) => assignmentString("indexes", v),
    layout_type: (v: object) => assignmentString("layout_type", v),
    legend_columns: (v: object) => assignmentString("legend_columns", v),
    legend_layout: (v: object) => assignmentString("legend_layout", v),
    legend_size: (v: object) => assignmentString("legend_size", v),
    live_span: (v: object) => assignmentString("live_span", v),
    logset: (_: any) => "", // 2.23.0 deprecated, see docs for widget.log_stream_definition
    margin: (v: object) => assignmentString("margin", v),
    markers: (v: object[]) => blockList(v, "marker", assignmentString),
    message_display: (v: object) => assignmentString("message_display", v),
    no_group_hosts: (v: object) => assignmentString("no_group_hosts", v),
    no_metric_hosts: (v: object) => assignmentString("no_metric_hosts", v),
    node_type: (v: object) => assignmentString("node_type", v),
    precision: (v: object) => assignmentString("precision", v),
    query: (v: object) => assignmentString("query", v),
    requests: (v: object) => convertRequests(v),
    right_yaxis: (v: object) => block("right_yaxis", v, assignmentString),
    scope: (v: object) => assignmentString("scope", v),
    service: (v: object) => assignmentString("service", v),
    show_breakdown: (v: object) => assignmentString("show_breakdown", v),
    show_date_column: (v: object) => assignmentString("show_date_column", v),
    show_distribution: (v: object) => assignmentString("show_distribution", v),
    show_error_budget: (v: object) => assignmentString("show_error_budget", v),
    show_errors: (v: object) => assignmentString("show_errors", v),
    show_hits: (v: object) => assignmentString("show_hits", v),
    show_last_triggered: (v: object) => assignmentString("show_last_triggered", v),
    show_latency: (v: object) => assignmentString("show_latency", v),
    show_legend: (v: object) => assignmentString("show_legend", v),
    show_message_column: (v: object) => assignmentString("show_message_column", v),
    show_resource_list: (v: object) => assignmentString("show_resource_list", v),
    show_tick: (v: object) => assignmentString("show_tick", v),
    size_format: (v: object) => assignmentString("size_format", v),
    sizing: (v: object) => assignmentString("sizing", v),
    slo_id: (v: object) => assignmentString("slo_id", v),
    sort: (v: object) => convertSort(v),
    span_name: (v: object) => assignmentString("span_name", v),
    start: (_: any) => "", // 2.23.0 deprecated, see docs for widget.manage_status_definition
    style: (v: object) => block("style", v, assignmentString),
    summary_type: (v: object) => assignmentString("summary_type", v),
    tags: (v: object) => assignmentString("tags", v),
    tags_execution: (v: object) => assignmentString("tags_execution", v),
    text: (v: object) => assignmentString("text", v),
    text_align: (v: object) => assignmentString("text_align", v),
    tick_edge: (v: object) => assignmentString("tick_edge", v),
    tick_pos: (v: object) => assignmentString("tick_pos", v),
    time: (v: any) => (!!v.live_span ? assignmentString("live_span", v.live_span) : ""),
    time_windows: (v: object) => assignmentString("time_windows", v),
    title: (v: object) => assignmentString("title", v),
    title_align: (v: object) => assignmentString("title_align", v),
    title_size: (v: object) => assignmentString("title_size", v),
    type: (_: any) => "",
    unit: (v: object) => assignmentString("unit", v),
    url: (v: object) => assignmentString("url", v),
    vertical_align: (_: any) => "", // 2.23.0 not described in docs, occurs in widget.note_definition json
    view_mode: (v: object) => assignmentString("view_mode", v),
    view_type: (v: object) => assignmentString("view_type", v),
    viz_type: (v: object) => assignmentString("viz_type", v),
    widget_layout: (v: object) => block("widget_layout", v, assignmentString),
    widgets: (v: object[]) => convertWidgets(v),
    xaxis: (v: object) => block("xaxis", v, assignmentString),
    yaxis: (v: object) => block("yaxis", v, assignmentString),
};

const REQUEST = {
    apm_query: (v: object) => assignmentString("apm_query", v),
    apm_stats_query: (v: object) => assignmentString("apm_stats_query", v),
    change_type: (v: object) => assignmentString("change_type", v),
    compare_to: (v: object) => assignmentString("compare_to", v),
    increase_good: (v: object) => assignmentString("increase_good", v),
    log_query: (v: object) =>
        block("log_query", v, (k1, v1) => convertFromDefinition(LOG_QUERY, k1, v1)),
    order_by: (v: object) => assignmentString("order_by", v),
    order_dir: (v: object) => assignmentString("order_dir", v),
    process_query: (v: object) => assignmentString("process_query", v),
    q: (v: object) => assignmentString("q", v),
    rum_query: (v: object) => assignmentString("rum_query", v),
    security_query: (v: object) => assignmentString("security_query", v),
    show_present: (v: object) => assignmentString("show_present", v),
    style: (v: object) => block("style", v, assignmentString),
    display_type: (v: object) => assignmentString("display_type", v),
    metadata: (v: object[]) => blockList(v, "metadata", assignmentString),
    network_query: (v: object) => assignmentString("network_query", v),
    on_right_yaxis: (v: object) => assignmentString("on_right_yaxis", v),
    aggregator: (v: object) => assignmentString("aggregator", v),
    alias: (v: object) => assignmentString("alias", v),
    cell_display_mode: (v: object) => assignmentString("cell_display_mode", v),
    conditional_formats: (v: object[]) => blockList(v, "conditional_formats", assignmentString),
    limit: (v: object) => assignmentString("limit", v),
    order: (v: object) => assignmentString("order", v),
    fill: (v: object) => block("fill", v, assignmentString),
};

const LOG_QUERY = {
    index: (v: object) => assignmentString("index", v),
    compute: (v: object) => block("compute_query", v, assignmentString),
    group_by: (v: object[]) => blockList(v, "group_by", (k1, v1) => convertFromDefinition(GROUP_BY, k1, v1)),
    multi_compute: (v: object[]) => blockList(v, "multi_compute", assignmentString),
    search: (v: any) => assignmentString("search_query", v.query),
    search_query: (v: object) => assignmentString("search_query", v),
};

const GROUP_BY = {
    facet: (v: object) => assignmentString("facet", v),
    limit: (v: object) => assignmentString("limit", v),
    sort: (v: object) => block("sort_query", v, assignmentString),
    sort_query: (v: object) => block("sort_query", v, assignmentString),
};

function convertSort(v: object) {
    return typeof v === "string"
        ? assignmentString("sort", v)
        : block("sort", v, assignmentString);
}

function convertWidgets(value: object[]) {
    return blockList(value, "widget", (k1, v1) => convertFromDefinition(WIDGET, k1, v1));
}

function convertRequests(value: object| object[]) {
    if (Array.isArray(value)) {
        return blockList(value, "request", (k, v) => convertFromDefinition(REQUEST, k, v));
    }
    return block("request", value, (k, v) => convertFromDefinition(REQUEST, k, v));
}

function widgetDefinition(contents: any) {
    let definitionType = contents.type === "slo" ? "service_level_objective" : contents.type;
    return block(`${definitionType}_definition`, contents, (k, v) =>
        convertFromDefinition(WIDGET_DEFINITION, k, v)
    );
}

export function generateDashboardTerraformCode(resourceName: string, dashboardData: object) {
    let result = "";
    Object.entries(dashboardData).forEach(([k, v]) => {
        result += convertFromDefinition(DASHBOARD, k, v);
    });
    return `resource "datadog_dashboard" "${resourceName}" {${result}\n}`;
}

File: cmd/dashboard_converter.go
package main

import (
	"fmt"
	"log"
	"reflect"
	"sort"
	"strings"
)

type (
	jmap       = map[string]any
	jmaps      = []jmap
	stringFunc func(any) string
)

var (
	blankGen = stringFunc(func(_ any) string { return "" })
)

func stringGen(name string) stringFunc {
	return func(v any) string { return assignmentString(name, v) }
}

var WIDGET_DEFINITION map[string]stringFunc

var DASHBOARD = map[string]stringFunc{
	"dashboard_lists_removed": blankGen,
	"description":             stringGen("description"),
	"id":                      blankGen,
	"is_read_only":            stringGen("is_read_only"),
	"layout_type":             stringGen("layout_type"),
	"notify_list":             stringGen("notify_list"),
	"reflow_type":             stringGen("reflow_type"),
	"restricted_roles":        func(_ any) string { return "// restricted_roles is still in beta" },
	"template_variables": func(v any) string {
		slice, ok := v.([]any)
		if !ok {
			log.Fatalf("template_variables expected as []any but got %T: %#v\n", v, v)
		}
		tvs := make(jmaps, len(slice))
		for i, tv := range slice {
			tvs[i], ok = tv.(jmap)
			if !ok {
				log.Fatalf("template_variables[%d] expected as jmaps but got %T: %#v\n", i, tv, tv)
			}
		}
		return blockList(tvs, "template_variable", assignmentString)
	},
	"template_variable_presets": func(v any) string {
		presets := must(jmapsFromAny(v))
		return blockList(presets, "template_variable_preset", func(k1 string, v1 any) string {
			return must(convertFromDefinition(TEMPLATE_VARIABLE_PRESET, k1, v1))
		})
	},
	"title": stringGen("title"),
	"url":   stringGen("url"),
	"widgets": func(v any) string {
		widgets := must(jmapsFromAny(v))
		return convertWidgets(widgets)
	},
}

var EVENT_QUERY = map[string]stringFunc{
	"aggregator":  stringGen("aggregator"),
	"compute":     func(v any) string { return block("compute", v.(jmap), assignmentString) },
	"data_source": stringGen("data_source"),
	"group_by": func(v any) string {
		groups := must(jmapsFromAny(v))
		return blockList(groups, "group_by", func(k1 string, v1 any) string {
			return must(convertFromDefinition(EVENT_QUERY_GROUP_BY, k1, v1))
		})
	},
	"indexes": stringGen("indexes"),
	"name":    stringGen("name"),
	"search":  func(v any) string { return block("search", v.(jmap), assignmentString) },
}

var EVENT_QUERY_GROUP_BY = map[string]stringFunc{
	"facet": stringGen("facet"),
	"limit": stringGen("limit"),
	"sort":  func(v any) string { return block("sort", v.(jmap), assignmentString) },
}

var FORMULA = map[string]stringFunc{
	"alias":   stringGen("alias"),
	"formula": stringGen("formula_expression "),
	"limit": func(v any) string {
		return block("limit", v.(jmap), func(k1 string, v1 any) string {
			return must(convertFromDefinition(FORMULA_LIMIT, k1, v1))
		})
	},
}

var FORMULA_LIMIT = map[string]stringFunc{
	"count": stringGen("count"),
	"order": stringGen("order"),
}

var GROUP_BY = map[string]stringFunc{
	"facet":      stringGen("facet"),
	"limit":      stringGen("limit"),
	"sort":       func(v any) string { return block("sort_query", v.(jmap), assignmentString) },
	"sort_query": func(v any) string { return block("sort_query", v.(jmap), assignmentString) },
}

var LOG_QUERY = map[string]stringFunc{
	"compute": func(v any) string {
		return block("compute_query", v.(jmap), assignmentString)
	},
	"group_by": func(v any) string {
		groups := must(jmapsFromAny(v))
		return blockList(groups, "group_by", func(k1 string, v1 any) string {
			return must(convertFromDefinition(GROUP_BY, k1, v1))
		})
	},
	"index": stringGen("index"),
	"multi_compute": func(v any) string {
		comps := must(jmapsFromAny(v))
		return blockList(comps, "multi_compute", assignmentString)
	},
	"search": func(v any) string {
		return assignmentString("search_query", v.(jmap)["query"])
	},
	"search_query": stringGen("search_query"),
}

var QUERY = map[string]stringFunc{
	"name":        stringGen("name"),
	"data_source": stringGen("data_source"),
	"query":       stringGen("query"),
}

var REQUEST = map[string]stringFunc{
	"aggregator":        stringGen("aggregator"),
	"alias":             stringGen("alias"),
	"apm_query":         stringGen("apm_query"),
	"apm_stats_query":   stringGen("apm_stats_query"),
	"cell_display_mode": stringGen("cell_display_mode"),
	"change_type":       stringGen("change_type"),
	"compare_to":        stringGen("compare_to"),
	"conditional_formats": func(v any) string {
		formats := must(jmapsFromAny(v))
		return blockList(formats, "conditional_formats", assignmentString)
	},
	"display_type": stringGen("display_type"),
	"fill":         func(v any) string { return block("fill", v.(jmap), assignmentString) },
	"formulas": func(v any) string {
		fs := must(jmapsFromAny(v))
		return blockList(fs, "formula", func(k1 string, v1 any) string {
			return must(convertFromDefinition(FORMULA, k1, v1))
		})
	},
	"increase_good": stringGen("increase_good"),
	"limit":         stringGen("limit"),
	"log_query": func(v any) string {
		return block("log_query", v.(jmap), func(k1 string, v1 any) string {
			return must(convertFromDefinition(LOG_QUERY, k1, v1))
		})
	},
	"metadata": func(v any) string {
		meta := must(jmapsFromAny(v))
		return blockList(meta, "metadata", assignmentString)
	},
	"network_query":  stringGen("network_query"),
	"on_right_yaxis": stringGen("on_right_yaxis"),
	"order":          stringGen("order"),
	"order_by":       stringGen("order_by"),
	"order_dir":      stringGen("order_dir"),
	"process_query":  stringGen("process_query"),
	"q":              stringGen("q"),
	"queries": func(v any) string {
		queries := must(jmapsFromAny(v))
		return queryBlockList(queries, assignmentString)
	},
	"response_format": stringGen("response_format"),
	"rum_query":       stringGen("rum_query"),
	"security_query":  stringGen("security_query"),
	"show_present":    stringGen("show_present"),
	"style": func(v any) string {
		return blockList(jmaps{v.(jmap)}, "style", assignmentString)
	},
}

var TEMPLATE_VARIABLE_PRESET = map[string]stringFunc{
	"name": stringGen("name"),
	"template_variables": func(v any) string {
		vars := must(jmapsFromAny(v))
		return blockList(vars, "template_variable", assignmentString)
	},
}

var WIDGET = map[string]stringFunc{
	"definition": func(v any) string { return widgetDefinition(v.(jmap)) },
	"id":         blankGen,
	"layout": func(v any) string {
		return block("widget_layout", v.(jmap), assignmentString)
	},
}

func init() {
	WIDGET_DEFINITION = map[string]stringFunc{
		"alert_id":         stringGen("alert_id"),
		"autoscale":        stringGen("autoscale"),
		"background_color": stringGen("background_color"),
		"check":            stringGen("check"),
		"color":            stringGen("color"),
		"color_by_groups":  stringGen("color_by_groups"),
		"color_preference": stringGen("color_preference"),
		"columns":          stringGen("columns"),
		"content":          stringGen("content"),
		"count":            blankGen,
		"custom_links": func(v any) string {
			links := must(jmapsFromAny(v))
			return blockList(links, "custom_link", assignmentString)
		},
		"custom_unit":    stringGen("custom_unit"),
		"display_format": stringGen("display_format"),
		"env":            stringGen("env"),
		"event":          func(v any) string { return block("event", v.(jmap), assignmentString) },
		"events": func(v any) string {
			events := must(jmapsFromAny(v))
			return blockList(events, "event", assignmentString)
		},
		"event_size":         stringGen("event_size"),
		"filters":            stringGen("filters"),
		"font_size":          stringGen("font_size"),
		"global_time_target": stringGen("global_time_target"),
		"group":              stringGen("group"),
		"group_by":           stringGen("group_by"),
		"grouping":           stringGen("grouping"),
		"has_padding":        blankGen,
		"has_search_bar":     stringGen("has_search_bar"),
		"hide_zero_counts":   stringGen("hide_zero_counts"),
		"indexes":            stringGen("indexes"),
		"layout_type":        stringGen("layout_type"),
		"legend_columns":     stringGen("legend_columns"),
		"legend_layout":      stringGen("legend_layout"),
		"legend_size":        stringGen("legend_size"),
		"live_span":          stringGen("live_span"),
		"logset":             blankGen,
		"margin":             stringGen("margin"),
		"markers": func(v any) string {
			markers := must(jmapsFromAny(v))
			return blockList(markers, "marker", assignmentString)
		},
		"message_display":     stringGen("message_display"),
		"no_group_hosts":      stringGen("no_group_hosts"),
		"no_metric_hosts":     stringGen("no_metric_hosts"),
		"node_type":           stringGen("node_type"),
		"precision":           stringGen("precision"),
		"query":               stringGen("query"),
		"requests":            convertRequests,
		"right_yaxis":         func(v any) string { return block("right_yaxis", v.(jmap), assignmentString) },
		"scope":               stringGen("scope"),
		"service":             stringGen("service"),
		"show_breakdown":      stringGen("show_breakdown"),
		"show_date_column":    stringGen("show_date_column"),
		"show_distribution":   stringGen("show_distribution"),
		"show_error_budget":   stringGen("show_error_budget"),
		"show_errors":         stringGen("show_errors"),
		"show_hits":           stringGen("show_hits"),
		"show_last_triggered": stringGen("show_last_triggered"),
		"show_latency":        stringGen("show_latency"),
		"show_legend":         stringGen("show_legend"),
		"show_message_column": stringGen("show_message_column"),
		"show_resource_list":  stringGen("show_resource_list"),
		"show_tick":           stringGen("show_tick"),
		"size_format":         stringGen("size_format"),
		"sizing":              stringGen("sizing"),
		"slo_id":              stringGen("slo_id"),
		"sort":                convertSort,
		"span_name":           stringGen("span_name"),
		"start":               blankGen,
		"style":               func(v any) string { return block("style", v.(jmap), assignmentString) },
		"summary_type":        stringGen("summary_type"),
		"tags":                stringGen("tags"),
		"tags_execution":      stringGen("tags_execution"),
		"text":                stringGen("text"),
		"text_align":          stringGen("text_align"),
		"tick_edge":           stringGen("tick_edge"),
		"tick_pos":            stringGen("tick_pos"),
		"time": func(v any) string {
			if liveSpan, ok := v.(jmap)["live_span"]; ok {
				return assignmentString("live_span", liveSpan)
			}
			return ""
		},
		"time_windows":   stringGen("time_windows"),
		"title":          stringGen("title"),
		"title_align":    stringGen("title_align"),
		"title_size":     stringGen("title_size"),
		"type":           blankGen,
		"unit":           stringGen("unit"),
		"url":            stringGen("url"),
		"vertical_align": blankGen,
		"view_mode":      stringGen("view_mode"),
		"view_type":      stringGen("view_type"),
		"viz_type":       stringGen("viz_type"),
		"widget_layout": func(v any) string {
			return block("widget_layout", v.(jmap), assignmentString)
		},
		"widgets": func(v any) string {
			return convertWidgets(must(jmapsFromAny(v)))
		},
		"xaxis": func(v any) string { return block("xaxis", v.(jmap), assignmentString) },
		"yaxis": func(v any) string { return block("yaxis", v.(jmap), assignmentString) },
	}
}

func convertEventQuery(value jmap) string {
	return block("query", value, func(_ string, _ any) string {
		return blockList(jmaps{value}, "event_query", func(k1 string, v1 any) string {
			return must(convertFromDefinition(EVENT_QUERY, k1, v1))
		})
	})
}

// convertRequests accepts either a single request as a jmap or a requests jmaps.
func convertRequests(value any) string {
	if reflect.ValueOf(value).Kind() == reflect.Slice {
		values := must(jmapsFromAny(value))
		return blockList(values, "request", func(k string, v any) string {
			return must(convertFromDefinition(REQUEST, k, v))
		})
	}
	return block("request", value.(jmap), func(k string, v any) string {
		return must(convertFromDefinition(REQUEST, k, v))
	})
}

func convertSort(v any) string {
	if sortStr, ok := v.(string); ok {
		return assignmentString("sort", sortStr)
	}
	return block("sort", v.(jmap), assignmentString)
}

func convertWidgets(value jmaps) string {
	return blockList(value, "widget", func(k1 string, v1 any) string {
		return must(convertFromDefinition(WIDGET, k1, v1))
	})
}

func widgetDefinition(contents jmap) string {
	definitionType := contents["type"].(string)
	if definitionType == "slo" {
		definitionType = "service_level_objective"
	}
	return block(fmt.Sprintf("\n%s_definition", definitionType), contents, func(k string, v any) string {
		return must(convertFromDefinition(WIDGET_DEFINITION, k, v))
	})
}

func generateDashboardTerraformCode(resourceName string, data jmap) (string, error) {
	var (
		result strings.Builder
		keys   = make([]string, 0, len(data))
	)
	for k := range data {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		s, err := convertFromDefinition(DASHBOARD, k, data[k])
		if err != nil {
			return "", err
		}
		result.WriteString(s)
	}
	return fmt.Sprintf("resource \"datadog_dashboard\" \"%s\" {%s\n}", resourceName, result.String()), nil
}
